esrally/driver/driver.py

```diff
@@ -313,42 +336,157 @@ def load_local_config(coordinator_config):
 
 
+class TrackPreparationActor(actor.RallyActor):
+    class Status(Enum):
+        INITIALIZING = "initializing"
+        PROCESSOR_RUNNING = "processor running"
+        PROCESSOR_COMPLETE = "processor complete"
+
+    def __init__(self):
+        super().__init__()
+        self.processors = queue.Queue()
+        self.original_sender = None
+        self.logger.info("Track Preparator started")
+        self.status = self.Status.INITIALIZING
+        self.children = []
+        self.tasks = []
+        self.cfg = None
+        self.data_root_dir = None
+        self.track = None
+
+
+    def receiveMsg_PoisonMessage(self, poisonmsg, sender):
+        self.logger.error("Track Preparator received a fatal indication from a load generator (%s). Shutting down.", poisonmsg.details)
+        self.send(self.original_sender, actor.BenchmarkFailure("Fatal track preparation indication", poisonmsg.details))
+
+    @actor.no_retry("track preparator")  # pylint: disable=no-value-for-parameter
+    def receiveMsg_ActorExitRequest(self, msg, sender):
+        self.logger.info("ActorExitRequest received. Forwarding to children")
+        for child in self.children:
+            self.send(child, msg)
+
+    @actor.no_retry("track preparator")  # pylint: disable=no-value-for-parameter
+    def receiveMsg_BenchmarkFailure(self, msg, sender):
+        # sent by our generic worker; give context and forward
+        msg.message = "Track preparator has detected a benchmark failure. Notifying master..."
+        self.send(self.original_sender, msg)
+
+    @actor.no_retry("track preparator")  # pylint: disable=no-value-for-parameter
+    def receiveMsg_PrepareTrack(self, msg, sender):
+        self.original_sender = sender
+        # load node-specific config to have correct paths available
+        self.cfg = load_local_config(msg.config)
+        self.data_root_dir = self.cfg.opts("benchmarks", "local.dataset.cache")
+        tpr = TrackProcessorRegistry(self.cfg)
+        self.track = msg.track
+
+        self.logger.info("Preparing track [%s]", self.track.name)
+        if self.track.has_plugins:
+            self.logger.info("Reloading track [%s] to ensure plugins are up-to-date.", self.track.name)
+            # the track might have been loaded on a different machine (the coordinator machine) so we force a track
+            # update to ensure we use the latest version of plugins.
+            load_track_plugins(self.cfg, self.track.name, register_track_processor=tpr.register_track_processor,
+                               force_update=True)
+
+        if tpr.processors():
+            # do this locally...
+            for processor in tpr.processors():
+                processor.on_after_load_track(self.track)
+            # ...and delegate the rest
+            self.children = [self._create_task_worker() for _ in range(num_cores(self.cfg))]
+            for processor in tpr.processors():
+                self.processors.put(processor)
```

Comment:
I am not certain why we can't enqueue all tasks upfront here? IMHO that would simplify the interaction model between this actor and its children. Every time we receive a `ReadyForWork` message we'd reply with another task until the queue is drained. We then need to reply with `TrackPrepared` to the original sender once all children have finished their outstanding work (for which we'd yet need to implement state handling logic).

An alternative would be to send all tasks round-robin to children, which would queue them themselves (in their thread pool's queue) and only respond when they're done with all submitted tasks. This would simplify state handling in this actor because we could rely on the base class implementation of `send_to_children_and_transition` and only need to send `TrackPrepared` from the callback method (which is invoked once all children have finished their tasks).

Reply:
calling this **The Grouping Comment** for reference elsewhere.

The logic here is to prevent overlap of different processors occurring at the same time.  Put another way, all given tasks will finish for a registered processor before starting the tasks for the next one.  I could pre-seed all the tasks here in this partitioned fashion, but that's not what you're talking about, so I'm guessing there would be no need (provided the way I implemented it to behave makes sense)